File: client-razor/Services/AuthService.cs
using client_razor.Models;
using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;

namespace client_razor.Services
{
    public class AuthService : IAuthService
    {
        private readonly HttpClient _httpClient;
        private readonly IHttpContextAccessor _httpContextAccessor;
        private readonly ILogger<AuthService> _logger;
        private const string TokenKey = "DenounceBeasts_Token";
        private const string UserKey = "DenounceBeasts_User";
        private const string ExpiryKey = "DenounceBeasts_TokenExpiry";

        public AuthService(
            HttpClient httpClient,
            IHttpContextAccessor httpContextAccessor,
            ILogger<AuthService> logger)
        {
            _httpClient = httpClient;
            _httpContextAccessor = httpContextAccessor;
            _logger = logger;

            // Set auth header if token exists
            var token = GetToken();
            if (!string.IsNullOrEmpty(token))
            {
                _httpClient.DefaultRequestHeaders.Authorization = 
                    new AuthenticationHeaderValue("Bearer", token);
            }
        }

        public async Task<AuthResponse?> LoginAsync(LoginRequest request)
        {
            try
            {
                _logger.LogInformation("Attempting login for user: {Email}", request.Email);

                var json = JsonSerializer.Serialize(request, new JsonSerializerOptions
                {
                    PropertyNamingPolicy = JsonNamingPolicy.CamelCase
                });

                var content = new StringContent(json, Encoding.UTF8, "application/json");
                var response = await _httpClient.PostAsync("api/auth/login", content);

                if (response.IsSuccessStatusCode)
                {
                    var responseContent = await response.Content.ReadAsStringAsync();
                    var authResponse = JsonSerializer.Deserialize<AuthResponse>(responseContent, new JsonSerializerOptions
                    {
                        PropertyNamingPolicy = JsonNamingPolicy.CamelCase
                    });

                    if (authResponse != null)
                    {
                        SetAuthData(authResponse.Token, authResponse.User, authResponse.Expires);
                        
                        // Set authorization header for subsequent requests
                        _httpClient.DefaultRequestHeaders.Authorization = 
                            new AuthenticationHeaderValue("Bearer", authResponse.Token);
                        
                        _logger.LogInformation("Login successful for user: {UserId}", authResponse.User.Id);
                        return authResponse;
                    }
                }

                var errorContent = await response.Content.ReadAsStringAsync();
                _logger.LogWarning("Login failed for user: {Email}. Status: {StatusCode}, Content: {Content}", 
                    request.Email, response.StatusCode, errorContent);
                
                return null;
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error during login for user: {Email}", request.Email);
                throw;
            }
        }

        public async Task<AuthResponse?> RegisterAsync(RegisterRequest request)
        {
            try
            {
                _logger.LogInformation("Attempting registration for user: {Email}", request.Email);

                var json = JsonSerializer.Serialize(request, new JsonSerializerOptions
                {
                    PropertyNamingPolicy = JsonNamingPolicy.CamelCase
                });

                var content = new StringContent(json, Encoding.UTF8, "application/json");
                var response = await _httpClient.PostAsync("api/auth/register", content);

                if (response.IsSuccessStatusCode)
                {
                    var responseContent = await response.Content.ReadAsStringAsync();
                    var authResponse = JsonSerializer.Deserialize<AuthResponse>(responseContent, new JsonSerializerOptions
                    {
                        PropertyNamingPolicy = JsonNamingPolicy.CamelCase
                    });

                    if (authResponse != null)
                    {
                        SetAuthData(authResponse.Token, authResponse.User, authResponse.Expires);
                        
                        // Set authorization header for subsequent requests
                        _httpClient.DefaultRequestHeaders.Authorization = 
                            new AuthenticationHeaderValue("Bearer", authResponse.Token);
                        
                        _logger.LogInformation("Registration successful for user: {UserId}", authResponse.User.Id);
                        return authResponse;
                    }
                }

                var errorContent = await response.Content.ReadAsStringAsync();
                _logger.LogWarning("Registration failed for user: {Email}. Status: {StatusCode}, Content: {Content}", 
                    request.Email, response.StatusCode, errorContent);
                
                return null;
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error during registration for user: {Email}", request.Email);
                throw;
            }
        }

        public Task LogoutAsync()
        {
            _logger.LogInformation("User logging out");
            
            ClearAuthData();
            
            // Clear authorization header
            _httpClient.DefaultRequestHeaders.Authorization = null;
            
            return Task.CompletedTask;
        }

        public async Task<AuthUser?> GetProfileAsync()
        {
            try
            {
                var response = await _httpClient.GetAsync("api/auth/profile");
                
                if (response.IsSuccessStatusCode)
                {
                    var responseContent = await response.Content.ReadAsStringAsync();
                    var user = JsonSerializer.Deserialize<AuthUser>(responseContent, new JsonSerializerOptions
                    {
                        PropertyNamingPolicy = JsonNamingPolicy.CamelCase
                    });

                    if (user != null)
                    {
                        // Update stored user info
                        var httpContext = _httpContextAccessor.HttpContext;
                        if (httpContext != null)
                        {
                            var userJson = JsonSerializer.Serialize(user);
                            httpContext.Session.SetString(UserKey, userJson);
                        }
                    }

                    return user;
                }

                _logger.LogWarning("Failed to get user profile. Status: {StatusCode}", response.StatusCode);
                return null;
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error getting user profile");
                return null;
            }
        }

        public async Task<bool> ChangePasswordAsync(ChangePasswordRequest request)
        {
            try
            {
                var json = JsonSerializer.Serialize(request, new JsonSerializerOptions
                {
                    PropertyNamingPolicy = JsonNamingPolicy.CamelCase
                });

                var content = new StringContent(json, Encoding.UTF8, "application/json");
                var response = await _httpClient.PostAsync("api/auth/change-password", content);

                if (response.IsSuccessStatusCode)
                {
                    _logger.LogInformation("Password changed successfully");
                    return true;
                }

                var errorContent = await response.Content.ReadAsStringAsync();
                _logger.LogWarning("Failed to change password. Status: {StatusCode}, Content: {Content}", 
                    response.StatusCode, errorContent);
                
                return false;
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error changing password");
                return false;
            }
        }

        public async Task<bool> IsEmailAvailableAsync(string email)
        {
            try
            {
                var response = await _httpClient.GetAsync($"api/auth/check-email?email={Uri.EscapeDataString(email)}");
                
                if (response.IsSuccessStatusCode)
                {
                    var responseContent = await response.Content.ReadAsStringAsync();
                    var result = JsonSerializer.Deserialize<Dictionary<string, bool>>(responseContent, new JsonSerializerOptions
                    {
                        PropertyNamingPolicy = JsonNamingPolicy.CamelCase
                    });

                    return result != null && !result.GetValueOrDefault("exists", true);
                }

                return false;
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error checking email availability for: {Email}", email);
                return false;
            }
        }

        public async Task<bool> ValidateTokenAsync()
        {
            try
            {
                var response = await _httpClient.GetAsync("api/auth/verify-token");
                return response.IsSuccessStatusCode;
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error validating token");
                return false;
            }
        }

        public bool IsAuthenticated()
        {
            var token = GetToken();
            var user = GetCurrentUser();
            
            if (string.IsNullOrEmpty(token) || user == null)
            {
                return false;
            }

            // Check if token is expired
            var httpContext = _httpContextAccessor.HttpContext;
            if (httpContext != null)
            {
                var expiryStr = httpContext.Session.GetString(ExpiryKey);
                if (!string.IsNullOrEmpty(expiryStr) && DateTime.TryParse(expiryStr, out var expiry))
                {
                    if (expiry <= DateTime.UtcNow.AddMinutes(-5)) // 5 minute buffer
                    {
                        ClearAuthData();
                        return false;
                    }
                }
            }

            return true;
        }

        public AuthUser? GetCurrentUser()
        {
            var httpContext = _httpContextAccessor.HttpContext;
            if (httpContext == null) return null;

            var userJson = httpContext.Session.GetString(UserKey);
            if (string.IsNullOrEmpty(userJson)) return null;

            try
            {
                return JsonSerializer.Deserialize<AuthUser>(userJson);
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error deserializing current user");
                return null;
            }
        }

        public string? GetToken()
        {
            var httpContext = _httpContextAccessor.HttpContext;
            return httpContext?.Session.GetString(TokenKey);
        }

        public void SetAuthData(string token, AuthUser user, DateTime expiry)
        {
            var httpContext = _httpContextAccessor.HttpContext;
            if (httpContext == null) return;

            httpContext.Session.SetString(TokenKey, token);
            httpContext.Session.SetString(UserKey, JsonSerializer.Serialize(user));
            httpContext.Session.SetString(ExpiryKey, expiry.ToString("O"));
            
            _logger.LogInformation("Authentication data set for user: {UserId}", user.Id);
        }

        public void ClearAuthData()
        {
            var httpContext = _httpContextAccessor.HttpContext;
            if (httpContext == null) return;

            httpContext.Session.Remove(TokenKey);
            httpContext.Session.Remove(UserKey);
            httpContext.Session.Remove(ExpiryKey);
            
            _logger.LogInformation("Authentication data cleared");
        }
    }
}

File: client-ts/src/services/auth-service.ts
/**
 * TypeScript Authentication Service
 * Handles JWT authentication with type safety and advanced features
 */

import { ApiClient } from './api-client';
import { EventEmitter } from '../utils/event-emitter';
import { Logger } from '../utils/logger';
import { Cache } from '../utils/cache';
import { 
  LoginDto, 
  RegisterDto, 
  AuthResponseDto, 
  AuthUser, 
  ChangePasswordDto 
} from '../types/entities';

export interface AuthServiceConfig {
  baseURL: string;
  tokenKey: string;
  userKey: string;
  expiryKey: string;
  refreshTokenKey?: string;
}

export interface AuthHeaders {
  'Content-Type': string;
  'Authorization'?: string;
}

export interface AuthEventData {
  user: AuthUser;
  timestamp: string;
}

export interface AuthErrorEventData {
  error: string;
  timestamp: string;
}

export class AuthService {
  private readonly apiClient: ApiClient;
  private readonly logger: Logger;
  private readonly eventEmitter: EventEmitter;
  private readonly config: AuthServiceConfig;
  
  private currentUser: AuthUser | null = null;
  private token: string | null = null;
  private tokenExpiry: Date | null = null;
  private refreshToken: string | null = null;
  private refreshTimer: number | null = null;

  constructor(config: Partial<AuthServiceConfig> = {}) {
    this.config = {
      baseURL: config.baseURL || 'https://localhost:7175',
      tokenKey: config.tokenKey || 'denounce_beasts_token',
      userKey: config.userKey || 'denounce_beasts_user',
      expiryKey: config.expiryKey || 'denounce_beasts_token_expiry',
      refreshTokenKey: config.refreshTokenKey || 'denounce_beasts_refresh_token'
    };

    this.apiClient = new ApiClient({ baseURL: this.config.baseURL });
    this.logger = new Logger('AuthService');
    this.eventEmitter = new EventEmitter();

    this._loadFromStorage();
    this._setupApiInterceptors();
    this._setupAutoRefresh();
  }

  /**
   * Initialize authentication service
   */
  public async initialize(): Promise<boolean> {
    try {
      if (this.isAuthenticated() && this._isTokenExpired()) {
        await this._attemptRefreshToken();
      }

      if (this.isAuthenticated()) {
        await this._validateCurrentToken();
      }

      this.logger.info('Authentication service initialized', {
        authenticated: this.isAuthenticated(),
        user: this.currentUser?.email
      });

      return this.isAuthenticated();
    } catch (error) {
      this.logger.error('Failed to initialize authentication service', error);
      this.logout();
      return false;
    }
  }

  /**
   * User login
   */
  public async login(loginData: LoginDto): Promise<AuthUser> {
    try {
      this.logger.info('Attempting login', { email: loginData.email });

      const response = await this.apiClient.post<AuthResponseDto>('/api/auth/login', loginData);

      if (!response.data) {
        throw new Error('Invalid response from server');
      }

      this._storeAuthData(response.data);

      this.eventEmitter.emit<AuthEventData>('login', {
        user: this.currentUser!,
        timestamp: new Date().toISOString()
      });

      this.logger.info('Login successful', {
        userId: this.currentUser!.id,
        email: this.currentUser!.email
      });

      return this.currentUser!;
    } catch (error) {
      this.logger.error('Login failed', error);
      
      this.eventEmitter.emit<AuthErrorEventData>('loginError', {
        error: error instanceof Error ? error.message : 'Login failed',
        timestamp: new Date().toISOString()
      });

      throw error;
    }
  }

  /**
   * User registration
   */
  public async register(registerData: RegisterDto): Promise<AuthUser> {
    try {
      this.logger.info('Attempting registration', { email: registerData.email });

      const response = await this.apiClient.post<AuthResponseDto>('/api/auth/register', registerData);

      if (!response.data) {
        throw new Error('Invalid response from server');
      }

      this._storeAuthData(response.data);

      this.eventEmitter.emit<AuthEventData>('register', {
        user: this.currentUser!,
        timestamp: new Date().toISOString()
      });

      this.logger.info('Registration successful', {
        userId: this.currentUser!.id,
        email: this.currentUser!.email
      });

      return this.currentUser!;
    } catch (error) {
      this.logger.error('Registration failed', error);
      
      this.eventEmitter.emit<AuthErrorEventData>('registerError', {
        error: error instanceof Error ? error.message : 'Registration failed',
        timestamp: new Date().toISOString()
      });

      throw error;
    }
  }

  /**
   * User logout
   */
  public logout(): void {
    this.logger.info('Logging out user', {
      userId: this.currentUser?.id,
      email: this.currentUser?.email
    });

    this._clearAuthData();

    this.eventEmitter.emit('logout', {
      timestamp: new Date().toISOString()
    });

    this.logger.info('Logout completed');
  }

  /**
   * Change user password
   */
  public async changePassword(changePasswordData: ChangePasswordDto): Promise<void> {
    try {
      this.logger.info('Attempting password change', {
        userId: this.currentUser?.id
      });

      await this.apiClient.post('/api/auth/change-password', changePasswordData);

      this.eventEmitter.emit('passwordChanged', {
        userId: this.currentUser!.id,
        timestamp: new Date().toISOString()
      });

      this.logger.info('Password changed successfully');
    } catch (error) {
      this.logger.error('Password change failed', error);
      throw error;
    }
  }

  /**
   * Get current user profile
   */
  public async getProfile(): Promise<AuthUser> {
    try {
      const response = await this.apiClient.get<AuthUser>('/api/auth/profile');
      
      if (response.data) {
        this.currentUser = response.data;
        this._saveToStorage();
      }

      return response.data!;
    } catch (error) {
      this.logger.error('Failed to get user profile', error);
      throw error;
    }
  }

  /**
   * Check if user is authenticated
   */
  public isAuthenticated(): boolean {
    return !!(this.token && this.currentUser && !this._isTokenExpired());
  }

  /**
   * Check if user has specific role
   */
  public hasRole(role: string): boolean {
    if (!this.isAuthenticated() || !this.currentUser) {
      return false;
    }
    return this.currentUser.roles.includes(role);
  }

  /**
   * Check if user has any of the specified roles
   */
  public hasAnyRole(roles: string[]): boolean {
    if (!this.isAuthenticated() || !this.currentUser) {
      return false;
    }
    return roles.some(role => this.hasRole(role));
  }

  /**
   * Check if user is admin
   */
  public isAdmin(): boolean {
    return this.hasRole('Administrador');
  }

  /**
   * Check if user is moderator or admin
   */
  public isModerator(): boolean {
    return this.hasAnyRole(['Moderador', 'Administrador']);
  }

  /**
   * Get current user
   */
  public getCurrentUser(): AuthUser | null {
    return this.currentUser;
  }

  /**
   * Get current token
   */
  public getToken(): string | null {
    return this.token;
  }

  /**
   * Get authentication headers
   */
  public getAuthHeaders(): AuthHeaders {
    const headers: AuthHeaders = {
      'Content-Type': 'application/json'
    };

    if (this.token) {
      headers['Authorization'] = `Bearer ${this.token}`;
    }

    return headers;
  }

  /**
   * Check if email exists
   */
  public async checkEmailExists(email: string): Promise<boolean> {
    try {
      const response = await this.apiClient.get<{ exists: boolean }>(
        `/api/auth/check-email?email=${encodeURIComponent(email)}`
      );
      return response.data?.exists || false;
    } catch (error) {
      this.logger.error('Failed to check email existence', error);
      return false;
    }
  }

  /**
   * Validate current token with server
   */
  private async _validateCurrentToken(): Promise<void> {
    try {
      await this.apiClient.get('/api/auth/verify-token');
    } catch (error) {
      this.logger.warn('Token validation failed', error);
      this.logout();
      throw error;
    }
  }

  /**
   * Store authentication data
   */
  private _storeAuthData(authData: AuthResponseDto): void {
    this.token = authData.token;
    this.currentUser = authData.user;
    this.tokenExpiry = new Date(authData.expires);

    this._saveToStorage();
    this._scheduleTokenRefresh();
  }

  /**
   * Save authentication data to localStorage
   */
  private _saveToStorage(): void {
    try {
      localStorage.setItem(this.config.tokenKey, this.token || '');
      localStorage.setItem(this.config.userKey, JSON.stringify(this.currentUser));
      localStorage.setItem(this.config.expiryKey, this.tokenExpiry?.toISOString() || '');
      
      if (this.refreshToken) {
        localStorage.setItem(this.config.refreshTokenKey!, this.refreshToken);
      }
    } catch (error) {
      this.logger.error('Failed to save auth data to storage', error);
    }
  }

  /**
   * Load authentication data from localStorage
   */
  private _loadFromStorage(): void {
    try {
      this.token = localStorage.getItem(this.config.tokenKey);
      this.refreshToken = localStorage.getItem(this.config.refreshTokenKey!);
      
      const userJson = localStorage.getItem(this.config.userKey);
      if (userJson) {
        this.currentUser = JSON.parse(userJson) as AuthUser;
      }

      const expiryString = localStorage.getItem(this.config.expiryKey);
      if (expiryString) {
        this.tokenExpiry = new Date(expiryString);
      }
    } catch (error) {
      this.logger.error('Failed to load auth data from storage', error);
      this._clearAuthData();
    }
  }

  /**
   * Clear all authentication data
   */
  private _clearAuthData(): void {
    this.token = null;
    this.currentUser = null;
    this.tokenExpiry = null;
    this.refreshToken = null;

    if (this.refreshTimer) {
      clearTimeout(this.refreshTimer);
      this.refreshTimer = null;
    }

    // Clear from storage
    localStorage.removeItem(this.config.tokenKey);
    localStorage.removeItem(this.config.userKey);
    localStorage.removeItem(this.config.expiryKey);
    localStorage.removeItem(this.config.refreshTokenKey!);
  }

  /**
   * Check if current token is expired
   */
  private _isTokenExpired(): boolean {
    if (!this.tokenExpiry) return true;

    // Add 5 minute buffer before expiry
    const bufferTime = 5 * 60 * 1000;
    return new Date().getTime() > (this.tokenExpiry.getTime() - bufferTime);
  }

  /**
   * Attempt to refresh token
   */
  private async _attemptRefreshToken(): Promise<void> {
    // For now, just logout when token expires
    // In a real implementation, you would call a refresh endpoint
    this.logger.info('Token expired, logging out user');
    this.logout();
  }

  /**
   * Schedule automatic token refresh
   */
  private _scheduleTokenRefresh(): void {
    if (this.refreshTimer) {
      clearTimeout(this.refreshTimer);
    }

    if (!this.tokenExpiry) return;

    // Schedule refresh 5 minutes before expiry
    const refreshTime = this.tokenExpiry.getTime() - Date.now() - (5 * 60 * 1000);
    
    if (refreshTime > 0) {
      this.refreshTimer = window.setTimeout(() => {
        this._attemptRefreshToken();
      }, refreshTime);
    }
  }

  /**
   * Setup automatic token refresh check
   */
  private _setupAutoRefresh(): void {
    // Check token validity every minute
    setInterval(() => {
      if (this.isAuthenticated() && this._isTokenExpired()) {
        this._attemptRefreshToken();
      }
    }, 60000);
  }

  /**
   * Setup API client interceptors
   */
  private _setupApiInterceptors(): void {
    // Request interceptor to add auth headers
    this.apiClient.addRequestInterceptor((config) => {
      if (this.token) {
        config.headers = config.headers || {};
        config.headers['Authorization'] = `Bearer ${this.token}`;
      }
      return config;
    });

    // Response interceptor to handle auth errors
    this.apiClient.addResponseInterceptor((response) => {
      if (response.status === 401 && this.isAuthenticated()) {
        this.logger.warn('Received 401 response, logging out user');
        this.logout();
      }
      return response;
    });
  }

  /**
   * Get event emitter for subscribing to auth events
   */
  public get events(): EventEmitter {
    return this.eventEmitter;
  }

  /**
   * Destroy the service and cleanup resources
   */
  public destroy(): void {
    if (this.refreshTimer) {
      clearTimeout(this.refreshTimer);
    }
    this.eventEmitter.removeAllListeners();
    this.logger.info('AuthService destroyed');
  }
}

// Create and export singleton instance
export const authService = new AuthService();